### ISSUE TITLE
Build on Illumos fails due to missing MAXPATHLEN

### ISSUE THREAD

**Issue by szaydel** [jmmv/kyua#97]
Should there be any expectation that the software should build on an Illumos-based OS? I have done some basic testing of configure/make and it fails during make, which does not surprise me. I am just wondering if anything has been done towards Illumos compatibility. I did not see much in commit logs.

**Comment by ngie-eign** [comment]
Can you please provide the make errors? There are probably some spots that need to be filled in with stub functions for returning system information/unmounting/etc.

**Comment by szaydel** [comment]
Thanks for a quick reply. Have a look, this is what I am observing. I have not had a chance to dig into this at all.

```
[root@box01 ~/kyua/kyua-testers-0.2]# gmake
gmake  all-am
gmake[1]: Entering directory '/root/kyua/kyua-testers-0.2'
gcc -DHAVE_CONFIG_H -I.  -DGDB=\"gdb\" -DKYUA_TMPDIR=\"/tmp\"   -g -O2  -D_FORTIFY_SOURCE=2 -Wall -Wcast-qual -Wextra -Wpointer-arith -Wredundant-decls -Wreturn-type -Wshadow -Wsign-compare -Wswitch -Wwrite-strings -DNDEBUG -Wmissing-prototypes -Wno-traditional -MT libtester_a-cli.o -MD -MP -MF .deps/libtester_a-cli.Tpo -c -o libtester_a-cli.o `test -f 'cli.c' || echo './'`cli.c
mv -f .deps/libtester_a-cli.Tpo .deps/libtester_a-cli.Po
gcc -DHAVE_CONFIG_H -I.  -DGDB=\"gdb\" -DKYUA_TMPDIR=\"/tmp\"   -g -O2  -D_FORTIFY_SOURCE=2 -Wall -Wcast-qual -Wextra -Wpointer-arith -Wredundant-decls -Wreturn-type -Wshadow -Wsign-compare -Wswitch -Wwrite-strings -DNDEBUG -Wmissing-prototypes -Wno-traditional -MT libtester_a-env.o -MD -MP -MF .deps/libtester_a-env.Tpo -c -o libtester_a-env.o `test -f 'env.c' || echo './'`env.c
mv -f .deps/libtester_a-env.Tpo .deps/libtester_a-env.Po
gcc -DHAVE_CONFIG_H -I.  -DGDB=\"gdb\" -DKYUA_TMPDIR=\"/tmp\"   -g -O2  -D_FORTIFY_SOURCE=2 -Wall -Wcast-qual -Wextra -Wpointer-arith -Wredundant-decls -Wreturn-type -Wshadow -Wsign-compare -Wswitch -Wwrite-strings -DNDEBUG -Wmissing-prototypes -Wno-traditional -MT libtester_a-error.o -MD -MP -MF .deps/libtester_a-error.Tpo -c -o libtester_a-error.o `test -f 'error.c' || echo './'`error.c
error.c: In function 'oom_format':
error.c:517:31: warning: unused parameter 'error' [-Wunused-parameter]
error.c: In function 'kyua_oom_error_new':
error.c:536:16: warning: unused variable 'ok' [-Wunused-variable]
mv -f .deps/libtester_a-error.Tpo .deps/libtester_a-error.Po
gcc -DHAVE_CONFIG_H -I.  -DGDB=\"gdb\" -DKYUA_TMPDIR=\"/tmp\"   -g -O2  -D_FORTIFY_SOURCE=2 -Wall -Wcast-qual -Wextra -Wpointer-arith -Wredundant-decls -Wreturn-type -Wshadow -Wsign-compare -Wswitch -Wwrite-strings -DNDEBUG -Wmissing-prototypes -Wno-traditional -MT libtester_a-fs.o -MD -MP -MF .deps/libtester_a-fs.Tpo -c -o libtester_a-fs.o `test -f 'fs.c' || echo './'`fs.c
fs.c: In function 'unmount_with_umount8':
fs.c:387:19: warning: unused variable 'ret' [-Wunused-variable]
fs.c: In function 'kyua_fs_current_path':
fs.c:488:34: error: expected expression before ':' token
Makefile:955: recipe for target 'libtester_a-fs.o' failed
gmake[1]: *** [libtester_a-fs.o] Error 1
gmake[1]: Leaving directory '/root/kyua/kyua-testers-0.2'
Makefile:613: recipe for target 'all' failed
gmake: *** [all] Error 2
[root@box01 ~/kyua/kyua-testers-0.2]#
```

**Comment by jmmv** [comment]
I do not "officially" intend to support Illumos right now because I do not have the time to deal with one more system... but certainly I will welcome portability fixes!

These warnings may be due to compiler differences in the handling of warnings; not uncommon. Give it another try with `./configure --disable-developer` to disable them and see if at least the software builds and runs fine.

**Comment by szaydel** [comment]
I can definitely try these couple of things to see if it is just a minor problem. Will report as soon as I had a chance to validate. Thank you both for feedback.

**Comment by jmmv** [comment]
Oh wow, yaneurabeya is right; copy/paste error. Stupid preprocessor conditionals.

There should be no extern "C" required here. kyua-testers is pure C and they don't offer any external library to link with C++.

**Comment by szaydel** [comment]
Maybe I am missing something, but I am only seeing MAXPATHLEN in one place, and that's where the issue is. Is the expectation that we'd get that from OS? I suppose I would assume that we want to get this from OS.

```
[root@box01 ~/kyua/kyua-testers-0.2]# gmake
gmake  all-am
gmake[1]: Entering directory '/root/kyua/kyua-testers-0.2'
gcc -DHAVE_CONFIG_H -I.  -DGDB=\"gdb\" -DKYUA_TMPDIR=\"/tmp\"   -g -O2  -D_FORTIFY_SOURCE=2 -Wall -Wcast-qual -Wextra -Wpointer-arith -Wredundant-decls -Wreturn-type -Wshadow -Wsign-compare -Wswitch -Wwrite-strings -DNDEBUG -Wmissing-prototypes -Wno-traditional -MT libtester_a-fs.o -MD -MP -MF .deps/libtester_a-fs.Tpo -c -o libtester_a-fs.o `test -f 'fs.c' || echo './'`fs.c
fs.c: In function 'unmount_with_umount8':
fs.c:387:19: warning: unused variable 'ret' [-Wunused-variable]
fs.c: In function 'kyua_fs_current_path':
fs.c:488:47: error: 'MAXPATHLEN' undeclared (first use in this function)
fs.c:488:47: note: each undeclared identifier is reported only once for each function it appears in
Makefile:955: recipe for target 'libtester_a-fs.o' failed
gmake[1]: *** [libtester_a-fs.o] Error 1
gmake[1]: Leaving directory '/root/kyua/kyua-testers-0.2'
Makefile:613: recipe for target 'all' failed
gmake: *** [all] Error 2
```

**Comment by szaydel** [comment]
So, I just patched fs.c with a `#define` for `MAXPATHLEN` setting it to `1024` and seems to have worked. I am thinking, there needs to be something that delivers a default value and overrides it if OS provided value differs.

```
[root@box01 ~/kyua/kyua-testers-0.2]# gmake
gmake  all-am
gmake[1]: Entering directory '/root/kyua/kyua-testers-0.2'
gcc -DHAVE_CONFIG_H -I.  -DGDB=\"gdb\" -DKYUA_TMPDIR=\"/tmp\"   -g -O2  -D_FORTIFY_SOURCE=2 -Wall -Wcast-qual -Wextra -Wpointer-arith -Wredundant-decls -Wreturn-type -Wshadow -Wsign-compare -Wswitch -Wwrite-strings -DNDEBUG -Wmissing-prototypes -Wno-traditional -MT libtester_a-fs.o -MD -MP -MF .deps/libtester_a-fs.Tpo -c -o libtester_a-fs.o `test -f 'fs.c' || echo './'`fs.c
fs.c: In function 'unmount_with_umount8':
fs.c:387:19: warning: unused variable 'ret' [-Wunused-variable]
mv -f .deps/libtester_a-fs.Tpo .deps/libtester_a-fs.Po
gcc -DHAVE_CONFIG_H -I.  -DGDB=\"gdb\" -DKYUA_TMPDIR=\"/tmp\"   -g -O2  -D_FORTIFY_SOURCE=2 -Wall -Wcast-qual -Wextra -Wpointer-arith -Wredundant-decls -Wreturn-type -Wshadow -Wsign-compare -Wswitch -Wwrite-strings -DNDEBUG -Wmissing-prototypes -Wno-traditional -MT libtester_a-result.o -MD -MP -MF .deps/libtester_a-result.Tpo -c -o libtester_a-result.o `test -f 'result.c' || echo './'`result.c
mv -f .deps/libtester_a-result.Tpo .deps/libtester_a-result.Po
gcc -DHAVE_CONFIG_H -I.  -DGDB=\"gdb\" -DKYUA_TMPDIR=\"/tmp\"   -g -O2  -D_FORTIFY_SOURCE=2 -Wall -Wcast-qual -Wextra -Wpointer-arith -Wredundant-decls -Wreturn-type -Wshadow -Wsign-compare -Wswitch -Wwrite-strings -DNDEBUG -Wmissing-prototypes -Wno-traditional -MT libtester_a-run.o -MD -MP -MF .deps/libtester_a-run.Tpo -c -o libtester_a-run.o `test -f 'run.c' || echo './'`run.c
run.c: In function 'mask_handlers':
run.c:115:15: warning: unused variable 'ret' [-Wunused-variable]
run.c: In function 'setup_signal':
run.c:212:15: warning: unused variable 'ret' [-Wunused-variable]
run.c: In function 'setup_timer':
run.c:230:15: warning: unused variable 'ret' [-Wunused-variable]
run.c: In function 'unlimit_core_size':
run.c:308:19: warning: unused variable 'ret' [-Wunused-variable]
run.c:312:15: warning: unused variable 'ret' [-Wunused-variable]
run.c: In function 'kyua_run_wait':
run.c:470:17: warning: unused variable 'waited_pid' [-Wunused-variable]
mv -f .deps/libtester_a-run.Tpo .deps/libtester_a-run.Po
gcc -DHAVE_CONFIG_H -I.  -DGDB=\"gdb\" -DKYUA_TMPDIR=\"/tmp\"   -g -O2  -D_FORTIFY_SOURCE=2 -Wall -Wcast-qual -Wextra -Wpointer-arith -Wredundant-decls -Wreturn-type -Wshadow -Wsign-compare -Wswitch -Wwrite-strings -DNDEBUG -Wmissing-prototypes -Wno-traditional -MT libtester_a-stacktrace.o -MD -MP -MF .deps/libtester_a-stacktrace.Tpo -c -o libtester_a-stacktrace.o `test -f 'stacktrace.c' || echo './'`stacktrace.c
stacktrace.c: In function 'run_gdb':
stacktrace.c:129:15: warning: unused variable 'input_fd' [-Wunused-variable]
mv -f .deps/libtester_a-stacktrace.Tpo .deps/libtester_a-stacktrace.Po
gcc -DHAVE_CONFIG_H -I.  -DGDB=\"gdb\" -DKYUA_TMPDIR=\"/tmp\"   -g -O2  -D_FORTIFY_SOURCE=2 -Wall -Wcast-qual -Wextra -Wpointer-arith -Wredundant-decls -Wreturn-type -Wshadow -Wsign-compare -Wswitch -Wwrite-strings -DNDEBUG -Wmissing-prototypes -Wno-traditional -MT libtester_a-text.o -MD -MP -MF .deps/libtester_a-text.Tpo -c -o libtester_a-text.o `test -f 'text.c' || echo './'`text.c
mv -f .deps/libtester_a-text.Tpo .deps/libtester_a-text.Po
rm -f libtester.a
ar cru libtester.a libtester_a-cli.o libtester_a-env.o libtester_a-error.o libtester_a-fs.o libtester_a-result.o libtester_a-run.o libtester_a-stacktrace.o libtester_a-text.o
ranlib libtester.a
depbase=`echo atf_list.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
gcc -DHAVE_CONFIG_H -I.     -g -O2  -D_FORTIFY_SOURCE=2 -Wall -Wcast-qual -Wextra -Wpointer-arith -Wredundant-decls -Wreturn-type -Wshadow -Wsign-compare -Wswitch -Wwrite-strings -DNDEBUG -Wmissing-prototypes -Wno-traditional -MT atf_list.o -MD -MP -MF $depbase.Tpo -c -o atf_list.o atf_list.c &&\
mv -f $depbase.Tpo $depbase.Po
depbase=`echo atf_result.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
gcc -DHAVE_CONFIG_H -I.     -g -O2  -D_FORTIFY_SOURCE=2 -Wall -Wcast-qual -Wextra -Wpointer-arith -Wredundant-decls -Wreturn-type -Wshadow -Wsign-compare -Wswitch -Wwrite-strings -DNDEBUG -Wmissing-prototypes -Wno-traditional -MT atf_result.o -MD -MP -MF $depbase.Tpo -c -o atf_result.o atf_result.c &&\
mv -f $depbase.Tpo $depbase.Po
atf_result.c: In function 'convert_result':
atf_result.c:682:1: warning: control reaches end of non-void function [-Wreturn-type]
rm -f libatf_tester.a
ar cru libatf_tester.a atf_list.o atf_result.o
ranlib libatf_tester.a
depbase=`echo tap_parser.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
gcc -DHAVE_CONFIG_H -I.     -g -O2  -D_FORTIFY_SOURCE=2 -Wall -Wcast-qual -Wextra -Wpointer-arith -Wredundant-decls -Wreturn-type -Wshadow -Wsign-compare -Wswitch -Wwrite-strings -DNDEBUG -Wmissing-prototypes -Wno-traditional -MT tap_parser.o -MD -MP -MF $depbase.Tpo -c -o tap_parser.o tap_parser.c &&\
mv -f $depbase.Tpo $depbase.Po
rm -f libtap_tester.a
ar cru libtap_tester.a tap_parser.o
ranlib libtap_tester.a
depbase=`echo atf_main.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
gcc -DHAVE_CONFIG_H -I.     -g -O2  -D_FORTIFY_SOURCE=2 -Wall -Wcast-qual -Wextra -Wpointer-arith -Wredundant-decls -Wreturn-type -Wshadow -Wsign-compare -Wswitch -Wwrite-strings -DNDEBUG -Wmissing-prototypes -Wno-traditional -MT atf_main.o -MD -MP -MF $depbase.Tpo -c -o atf_main.o atf_main.c &&\
mv -f $depbase.Tpo $depbase.Po
gcc  -g -O2  -D_FORTIFY_SOURCE=2 -Wall -Wcast-qual -Wextra -Wpointer-arith -Wredundant-decls -Wreturn-type -Wshadow -Wsign-compare -Wswitch -Wwrite-strings -DNDEBUG -Wmissing-prototypes -Wno-traditional   -o kyua-atf-tester atf_main.o libatf_tester.a libtester.a
depbase=`echo plain_main.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
gcc -DHAVE_CONFIG_H -I.     -g -O2  -D_FORTIFY_SOURCE=2 -Wall -Wcast-qual -Wextra -Wpointer-arith -Wredundant-decls -Wreturn-type -Wshadow -Wsign-compare -Wswitch -Wwrite-strings -DNDEBUG -Wmissing-prototypes -Wno-traditional -MT plain_main.o -MD -MP -MF $depbase.Tpo -c -o plain_main.o plain_main.c &&\
mv -f $depbase.Tpo $depbase.Po
gcc  -g -O2  -D_FORTIFY_SOURCE=2 -Wall -Wcast-qual -Wextra -Wpointer-arith -Wredundant-decls -Wreturn-type -Wshadow -Wsign-compare -Wswitch -Wwrite-strings -DNDEBUG -Wmissing-prototypes -Wno-traditional   -o kyua-plain-tester plain_main.o libtester.a
depbase=`echo tap_main.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
gcc -DHAVE_CONFIG_H -I.     -g -O2  -D_FORTIFY_SOURCE=2 -Wall -Wcast-qual -Wextra -Wpointer-arith -Wredundant-decls -Wreturn-type -Wshadow -Wsign-compare -Wswitch -Wwrite-strings -DNDEBUG -Wmissing-prototypes -Wno-traditional -MT tap_main.o -MD -MP -MF $depbase.Tpo -c -o tap_main.o tap_main.c &&\
mv -f $depbase.Tpo $depbase.Po
gcc  -g -O2  -D_FORTIFY_SOURCE=2 -Wall -Wcast-qual -Wextra -Wpointer-arith -Wredundant-decls -Wreturn-type -Wshadow -Wsign-compare -Wswitch -Wwrite-strings -DNDEBUG -Wmissing-prototypes -Wno-traditional   -o kyua-tap-tester tap_main.o libtap_tester.a libtester.a
name=kyua-atf-tester.1; sed -e 's,__PACKAGE__,kyua-testers,g' -e 's,__TMPDIR__,/tmp,g' <./${name}.in >${name}.tmp; mv ${name}.tmp ${name}
name=kyua-plain-tester.1; sed -e 's,__PACKAGE__,kyua-testers,g' -e 's,__TMPDIR__,/tmp,g' <./${name}.in >${name}.tmp; mv ${name}.tmp ${name}
name=kyua-tap-tester.1; sed -e 's,__PACKAGE__,kyua-testers,g' -e 's,__TMPDIR__,/tmp,g' <./${name}.in >${name}.tmp; mv ${name}.tmp ${name}
name=kyua-tester.1; sed -e 's,__PACKAGE__,kyua-testers,g' -e 's,__TMPDIR__,/tmp,g' <./${name}.in >${name}.tmp; mv ${name}.tmp ${name}
sed -e 's#__LIBEXECDIR__#/usr/local/libexec#g' \
    -e 's#__VERSION__#0.2#g' \
    <./kyua-testers.pc.in >kyua-testers.pc.tmp
mv kyua-testers.pc.tmp kyua-testers.pc
gmake[1]: Leaving directory '/root/kyua/kyua-testers-0.2'
[root@box01 ~/kyua/kyua-testers-0.2]# echo $?
0
```

**Comment by jmmv** [comment]
`MAXPATHLEN` is typically provided by the operating system in `sys/param.h`. The only system I knew to not implement this was The Hurd as they refuse to have fixed-length buffers. Try a `grep -r MAXPATHLEN /usr/include` and see what shows up.

If the constant is nowhere to be found, we'll have to add a `configure.ac` check to test for its presence and define a fake value in `config.h` otherwise.

**Comment by jmmv** [comment]
Please reopen when you provide feedback. I would like to know if `MAXPATHLEN` is indeed missing from Illumos or we are just failing to include some specific header file.

**Comment by rodrigc** [comment]
In the Illumos community, there are multiple distributions.  Each distribution uses a different packaging infrastructure.
Joyent's Smartos uses pkgsrc, and kyua is failing to build there:

http://us-east.manta.joyent.com/pkgsrc/public/reports/upstream-trunk32/20140907.2109/kyua-0.10/build.log

**Comment by rodrigc** [comment]
I setup a VM of SmartOS which is an Illumos distribution: http://smartos.org/

I then followed these instructions to set up pkgsrc under Illumos:

http://pkgsrc.joyent.com/installing.html#installing-on-illumos

I then did cd /opt/pkgsrc/devel/kyua; bmake

and was able to reproduce the failure.  Illumos _does_ have MAXPATHLEN
in <sys/param.h>.  On Illumos, it doesn't get sucked in the same way as on FreeBSD and other
platforms, so it needs to be explicitly included.  There are other minor compilation failures,
so I will submit a patch.
